daemon/src/order/taker.rs

```diff
@@ -0,0 +1,306 @@
+                    leverage,
+                    settlement_interval,
+                    Role::Taker,
+                    quantity,
+                    // Completely irrelevant when using libp2p
+                    Identity::new(x25519_dalek::PublicKey::from(
+                        *b"hello world, oh what a beautiful",
+                    )),
+                    Some(maker.into()),
+                    opening_fee,
+                    funding_rate,
+                    tx_fee_rate,
+                );
+
+                // If this fails we shouldn't try to append
+                // `ContractSetupFailed` to the nonexistent CFD
+                if let Err(e) = db.insert_cfd(&cfd).await {
+                    tracing::error!("Inserting new cfd failed: {e:#}");
+                    return anyhow::Ok(());
+                };
+
+                projection.send(projection::CfdChanged(cfd.id())).await?;
+
+                let stream = endpoint
+                    .send(OpenSubstream::single_protocol(maker, PROTOCOL_NAME))
+                    .await??;
+
+                let mut framed =
+                    Framed::new(stream, JsonCodec::<TakerMessage, MakerMessage>::new());
+
+                framed
+                    .send(TakerMessage::PlaceOrder {
+                        id,
+                        quantity,
+                        leverage,
+                        position,
+                        opening_price,
+                        settlement_interval,
+                        opening_fee,
+                        funding_rate,
+                        tx_fee_rate,
+                        oracle_event_id,
+                    })
+                    .await?;
+
+                match framed.next().await.context("Stream terminated")?? {
+                    MakerMessage::Decision(Decision::Accept) => {
+                        tracing::info!(order_id = %msg.id, peer = %maker, "Order accepted");
+                    }
+                    MakerMessage::Decision(Decision::Reject) => {
+                        tracing::info!(order_id = %msg.id, peer = %maker, "Order rejected");
+
+                        executor
+                            .execute(id, |cfd| {
+                                cfd.reject_contract_setup(anyhow::anyhow!("Unknown"))
+                            })
+                            .await?;
+
+                        return anyhow::Ok(());
+                    }
+                    MakerMessage::ContractSetupMsg(_) => bail!("Unexpected message"),
+                };
+
+                let (setup_params, position) = executor
+                    .execute(id, |cfd| cfd.start_contract_setup())
+                    .await?;
+
+                let (sink, stream) = framed.split();
+
+                let announcement = get_announcement
+                    .send(oracle::GetAnnouncement(oracle_event_id))
+                    .await??;
+
+                let dlc = setup_contract::new(
```

Comment:
I would only do it if it's necessary; if it works well without this, I see no point in doing so; I doubt that such a refactor would even be more maintainable...


> I assume that you are aware that the protocol changes will require logic changes to the maker's decision to evaluate if an `order` is acceptable.
> 
> We should evaluate implementing at least some of the acceptance criteria before we can roll this out. Unless you are already working on it please create a ticket for that. We have to make sure to do this before releasing.

great point. I will add it an additional commit here, because without this it's not a complete PR :)